improvements to integration that relies on meijerint

smichr opened the pull request:
fixes #8368 and closes #8429
- [x] decide on s + t > u (as in Prudnikov) or m + n > 0
##### decision is to not get into that with this PR; my `jk` branch has some experimentation and modifications

fixes #8901
fixes #7173

smichr commented:
ping @jksuom

jksuom commented on sympy/integrals/integrals.py:
```diff
@@ -436,6 +438,8 @@ def doit(self, **hints):
             def try_meijerg(function, xab):
                 ret = None
+                function = function.xreplace(dict([(h, h.rewrite(exp))
+                    for h in function.atoms(HyperbolicFunction)]))
```
Exponentials are certainly more simple to express as G-functions than hyperbolics. I wonder if there is any particular reason why this should be so early in integrals.py instead of meijerint.py.; for example, in _meijerint_definite_2 before the _guess_expansion -loop.

smichr replied:
I originally tried this in the `_guess_expansion` but some of the tests I added failed if it was done that late. I'll try again to see if there is a better place withing meijer itself.

jksuom commented on sympy/integrals/meijerint.py:
```diff
@@ -1807,20 +1782,23 @@ def _guess_expansion(f, x):
     from sympy import expand_trig
     from sympy.functions.elementary.trigonometric import TrigonometricFunction
-    from sympy.functions.elementary.hyperbolic import HyperbolicFunction
     res = [(f, 'originial integrand')]
```
original?

smichr replied:
Indeed! Thanks.

jksuom commented on sympy/integrals/heurisch.py:
```diff
@@ -558,8 +558,20 @@ def find_non_syms(expr):
-            solution = [ (k.as_expr(), v.as_expr()) for k, v in solution.items() ]
-            return candidate.subs(solution).subs(list(zip(coeffs, [S.Zero]*len(coeffs))))
+            sol = {}
+            # remove coeff from k in case Floats were involved
```
With AlgebraicField the situation is even worse. Its dtype is ANP which has no attribute 'as_expr'.

jksuom commented on sympy/integrals/heurisch.py:
```diff
@@ -558,8 +558,20 @@ def find_non_syms(expr):
+                if not k.is_Symbol:
+                    if k.is_Mul:
+                        c, k = k.as_coeff_Mul()
+                        v /= c
```
The way solve_lin_sys() works, it looks like c will always be the unit of the field.

smichr replied:
In that case, perhaps the change can happen (and I would prefer that it do) in `solve_lin_sys`. I would be open to any suggestions from you, perhaps as PRs to what I have here.

I was able to move the rewriting into the `meijerint*` routines directly.

jksuom replied:
I was also beginning to think that `solve_lin_sys()` should be changed. In fact, it is natural to expect that the keys were symbols.

jksuom replied:
Before I forget: Here are a couple of lines in meijerint.py:1870 that should be rewritten.
```
        ress = [_meijerint_definite_4(g, x) for g in f.args]
        if all(r is not None for r in ress):
```

When studying 8882 I met an Add with 65 arguments for each one of which `_meijerint_definite_4` returns None. The code diligently fills `ress` with 65 Nones wasting a long time even if one call only takes seconds. I am not quite sure which would be the preferred way to make the code return at the first None.

jksuom replied:
I have been thinking how to change solve_lin_sys though it is too late to do that tonight. It implies practically reverting your changes to heurisch.py. Is it convenient to work on this branch or should I create a new one?

smichr committed:
solve: watch for expressions which simplify to a constant

smichr committed:
add relatively high-level checks of nth args

Similar checks could go in the low level routines in polyclasses or
densebasic, e.g.

          def nth(f, *N):
              """Returns the ``n``-th coefficient of ``f``. """
              if all(isinstance(n, int) for n in N):
                  # check validity of N
     +            ngens = 0
     +            l = f.rep
     +            while type(l) is list:
     +                ngens += 1
     +                l = l[0]
     +            if len(N) != ngens:
     +                raise ValueError
                  return dmp_ground_nth(f.rep, N, f.lev,
                      f.dom)

But such a change was not made since those are such low-level routines.

smichr committed:
8792: minor modifications to meijerint

o correct variables in loop (issue 8792)

o convert loops to list comprehensions and allow c1 to get a False quicker

o an exit is added to loops in which res is being computed if the condition
  will be False; this follows the pattern that was established a similar
  piece of code in meijerint_inversion

smichr committed:
8368: rewrite hyperbolics in terms of exp for meijerint*

Issue 8368 demonstrates that meijerint has problems with hyperbolics, so
they are now rewritten in terms of exp. With this change, _gues_expansion
no longer needs to watch for hyperbolic functions to expand.

Also, redundant expansions are not returned from _guess_expansion.

smichr committed:
meijerint: use hyperbolic rewrite as last resort

jksuom commented on sympy/integrals/tests/test_meijerint.py:
```diff
@@ -631,5 +631,14 @@ def test_fresnel():
+        -t*exp(2*x)/(2*t**2*exp(x)*exp(t*x) - 2*exp(x)*exp(t*x)) -
+        t/(2*t**2*exp(x)*exp(t*x) - 2*exp(x)*exp(t*x)) -
+        exp(2*x)/(2*t**2*exp(x)*exp(t*x) - 2*exp(x)*exp(t*x)) +
+        1/(2*t**2*exp(x)*exp(t*x) - 2*exp(x)*exp(t*x)))
```
Perhaps simplify() would be called for here.

smichr replied:
I tried something judicious for those returns obtained after rewriting in terms of exp.

smichr committed:
modify test for non-deterministic result

smichr committed:
make heurish deterministic wrt ground

smichr committed:
Revert "make heurish deterministic wrt ground"

This reverts commit a376a303ea30a352334ede054c4ad13a50dc44ab.

The problem only appeared to go away because a different hash was
needed to produce the error.

smichr committed:
7173: use positive instead of > to check assumption

jksuom committed:
Make solve_lin_sys return symbols as keys

Conflicts:
	sympy/integrals/heurisch.py

smichr commented:
I just wanted to try it both ways...it seems like the hyperbolic rewrite is needed less often with the "> 0" condition, but I'll stick with what is in the book and reset the last commit.

jksuom commented:
I have modified solve_lin_sys to return symbols as keys. It is at github/jksuom:8368. (I have yet to learn how to send a PR...)

smichr commented:
I can check that PR out in a bit. SOmething interesting about the s + t > u instead of s + t > 0: now the lapalce_transform of Chi(x) fails:

```
>>> laplace_transform(Chi(x), x, s)
LaplaceTransform(Chi(x), x, s)
>>> laplace_transform(Chi(x).rewrite(expint), x, s)
(-(log(s - 1) + log(s + 1))/(2*s), 1, True)
```

smichr commented:
> I have yet to learn how to send a PR...

I cherry picked this, but for future reference: you start the same was as when you create a regular PR except that you change the **base fork** to, for example, smichr/sympy, and then select the **branch** that you want to send the PR to. The only catch is that not everyone shows up in the base fork list. What I generally do is just select someone then update the url, changing 'someone:branch' to 'right-person:right-branch'.

jksuom commented:
Thanks, I'll try that next time.

jksuom commented:
The main part of Chi(x) is the integral of  (cosh(x) - 1)/x,  which does not seem so bad. But
```
In [23]: laplace_transform((cosh(x) - 1)/x, x, s)
...
NotImplementedError: Result depends on the sign of sign(polar_lift(-_s + 1))
```

smichr commented:
Also, the `s + t > u` caused something to hang in `test_mellin_transform_bessel`.

jksuom commented:
I'll look into the `s + t > u`  case tomorrow. How do I check that out?

smichr committed:
solve_lin_sys gets option to return symbols

smichr commented:
> How do I check that out?

Assuming you have pulled my branch in to your computer, you can then check out any commit by seeing what the associated SHA1 is for it. You can use the shortened form from `git log --oneline`. The commit that makes s + t > u has a SHA1 of  67ba782 so you just `git checkout  67ba782`. If you make changes that you want to keep you can then commit them and create a new branch, e.g. `git checkout -b newbranch`; if you don't do that then when you checkout anything else, the changes will be lost.

smichr commented:
I changed a commit and the SHA1s changed so the pertinent commit is now at 67ed26c...but whatever I do here, just check with `git log --oneline` and you can see the SHA1 for the commit you are interested in.

Thanks for looking into this. Things really should be symmetric wrt cosh and sinh, I would think. And both sinh and cosh should work (I'm guessing) without the rewrite option. Why that is not happening is the question at hand.

jksuom commented:
I have tried to analyze the hang in `test_mellin_transform_bessel` with `s + t > u`. It seems that there are two problems. `meijerint_definite` decides very quickly that it cannot do the integral. I suspect that there is a bug that may not be connected with the inequalities. `integrate` then falls back to heurisch. Contrary to what the docstring says it cannot deal with Bessel functions in its present form.

The problem is the following: heurisch computes the derivatives of the functions. For example
```
In [20]: print(besseli(a, x).diff(x))
besseli(a - 1, x)/2 + besseli(a + 1, x)/2
```
It then computes the derivatives of the new terms obtaining terms with parameters  `a - 2`  and  `a + 2`, etc. This leads to a great number of variables and polynomials of high degree. As a result the memory appears to fill so that the process effectively comes stop. 

I think there is a way to solve this problem but not quite easily. The quick and dirty interim solution would be to return as soon as a Bessel function is met in heurisch.

As to the suspected bug in meijerint, I intend to examine that in more detail. But it will still take some time.

smichr committed:
always sort mapping in heurisch

smichr committed:
make heurisch more deterministic

smichr commented:
Thanks again @jksuom for looking at that. I wish I could watch over your shoulder as you work! If it's ok with you I will commit what is here, leaving meijerint conditions alone. I pushed to my repo a branch named jk that has the different variations of s + t and m + n in it if you want to look at that or copy to use for your own testing.

jksuom commented:
I also think it is time to merge a somewhat stable result. I am still going to continue with meijerint hoping to better understand the various conditions for the integrals.

smichr commented:
I suspect there is still something non-deterministic about the heurisch algorithm. There was a failure in `test_hyperexpand/test_prudnikov_11` that I saw in a Travis test but have not been able to reproduce after 400 reruns here and 5 travis runs. But master is less deterministic, so I am committing what is here.

skirpichev commented:
@asmeurer, could you review heurisch changes here?

Esp. last commit, 2e06e1a.  I can't reproduce failure in test_issue_8901 after 50 rerun's (nor after removing 073b0e6).

But if that make sense, shouldn't monoms be ordered like irreducibles in the last commit?  See https://github.com/skirpichev/omg/pull/57

asmeurer commented:
That change seems innocuous. It's probably just masking some nondeterminism somewhere else, since that's only determining what order the dummy symbols are named in.
